tests/python/pants_test/util/test_memo.py

```diff
@@ -354,3 +358,48 @@ def calls(self):
 
     self.assertEqual(4, foo2.calls)
     self.assertEqual(4, foo2.calls)
+
+  def test_collection_coercion(self):
+    @memoized
+    def f(x):
+      return sum(x)
+    with self.assertRaisesRegexp(TypeError, re.escape("unhashable type: 'list'")):
+      f([3, 4])
+
+    g_called = self._Called(increment=1)
+    @memoized(key_factory=coerce_collections)
+    def g(x):
+      g_called._called()
+      return sum(x)
+    x = [3, 4]
+    # x is converted into a tuple by coerce_collections, so this will only call g once.
+    self.assertEqual(7, g(tuple(x)))
+    self.assertEqual(7, g(x))
+    x[0] = 2
+    self.assertEqual(6, g(x))
+    # OrderedSet is converted into a tuple which is equal to the previous call, so this should not
+    # increase the call count.
+    self.assertEqual(6, g(OrderedSet(x)))
+    self.assertEqual(2, g_called._calls)
+
+    class C(self._Called):
```

Comment:
This should break out into its own test method, its testing a different function.